### PR TITLE
Add format debounce and improve debounce time setup

### PR DESCRIPTION
**Pull request by ray-x**
The PR
1) Trying to improve the lintDebounce setup.  Currently it only accepts int and it likely to mistake the time unit to ms. The golang Duration unit is ns.  That means the setup need to use 500000000 to debounce 500ms. I did a quick grep of setups (dotfiles) in github. Most people are setting this field to 500, I believe they mean to set up debounce to 500ms. The changes allow user setup the debounce to "500ms" or "0.5s".

2) Add format debounce. I saw random failures when multiple format tools were configured and will trigger lsp format multiple times. This sometime result in an incorrect result.

### CODE REVIEW

**Comment by mattn** [comment]
Please rebase master branch

File: langserver/handle_text_document_formatting.go [hunk 1]
```diff
@@ -11,11 +11,16 @@ import (
 	"regexp"
 	"runtime"
 	"strings"
+	"time"
 
 	"github.com/sourcegraph/jsonrpc2"
 )
 
-func (h *langHandler) handleTextDocumentFormatting(ctx context.Context, conn *jsonrpc2.Conn, req *jsonrpc2.Request) (result interface{}, err error) {
+func (h *langHandler) handleTextDocumentFormatting(
```
Comment (mattn) [inline langserver/handle_text_document_formatting.go:19]: I don't prefer this change.

**Comment by mattn** [comment]
>Currently it only accepts int 

Really?

File: langserver/handle_text_document_formatting.go (hunk 1 above)
Comment (ray-x) [reply langserver/handle_text_document_formatting.go]: This change was reverted.

**Comment by ray-x** [comment]
Hi, Matt
The branch was updated. 
Thanks!

**Comment by mattn** [comment]
I think that Duration is not needed.

**Comment by ray-x** [comment]
The change to Duration is necessary
1) It allows users set the interval in the format of `1s`, `0.5s`, `30ms`. This is more explicit. 
2) Golang interval is nanosecond based. That is foreign to the user from lua/vimscript world where the setup is millisecond based. By searching the github I found most of users set the interval to values `200`. But it won't work as that is 0.0000002ms. So it would be better to specify the setup in format of `200ms`. And `Duration` still allow set value in float format for backward capacity.  Without this change, the setup of `200ms` will fail to parse and fall back to 0. 
There are discussions https://github.com/golang/go/issues/10275

**Comment by mattn** [comment]
go-yaml seems to be already support time.Duration.

https://github.com/go-yaml/yaml/blob/7649d4548cb53a614db133b2a8ac1f31859dda8c/decode_test.go#L542

**Comment by ray-x** [comment]
I did not test yaml setup for efm-languageserver
The reason for the change is neovim is using json to communicate to efm.  that is not covered by go-yaml package.

**Comment by mattn** [comment]
Could you please show me your neovim configuration for the efm-langserver?

**Comment by ray-x** [comment]
```lua
  lspconfig.efm.setup {
    flags = {debounce_text_changes = 1000},
    cmd = {'efm-langserver', '-loglevel', '2', '-logfile', '/tmp/efm.log'}, 
    init_options = {documentFormatting = true},
    on_attach = function(client)
      client.resolved_capabilities.document_formatting = true
      client.resolved_capabilities.goto_definition = false
      client.resolved_capabilities.code_action = nil
      local log = require("guihua.log").new({level = "info"}, true)
      vim.cmd([[autocmd BufWritePre <buffer> lua vim.lsp.buf.formatting()]])
    end,
    filetypes = {
      "javascript", "javascriptreact", 'typescript', 'typescriptreact', 
      'html', 'css', 'go', 'lua'
    },
    settings = {
      rootMarkers = {".git/", 'package.json', 'Makefile', 'go.mod'},
      lintDebounce =   "2s",
      formatDebounce = "1000ms",
      languages = {
        typescript = {stylelint, prettier},
        typescriptreact = {stylelint, prettier},
        javascript = {eslint_d},
        javascriptreact = {eslint_d},
       python = { python-flake8 },
        go = {
          {
            formatCommand = "golines --max-len=120  --base-formatter=gofumpt",
            formatStdin = true,
            -- "staticcheck" ?
            lintCommand = "golangci-lint run",
            LintSeverity = 3,
          }
        },
        lua = {
          { formatCommand = "lua-format --indent-width 2 --tab-width 2 --no-use-tab --column-limit 120 --column-table-limit 100 --no-keep-simple-function-one-line --no-chop-down-table --chop-down-kv-table --no-keep-simple-control-block-one-line --no-keep-simple-function-one-line --no-break-after-functioncall-lp --no-break-after-operator",
           formatStdin = true,
          }
        },
      }
    }
  }

```

**Comment by mattn** [comment]
Your `"2s"` is converted to JSON value as string for workspace/didChangeConfiguration. And it is decoded as `json.Unmarshal`

https://github.com/mattn/efm-langserver/blob/4e23221382658949c836a11874af265179be2dee/langserver/handle_workspace_did_change_configuration.go#L16

As far as I can see, it should work correctly.

**Comment by ray-x** [comment]
In fact, I am seeing an error in that file (with the config I post earlier):

```
2021/07/07 16:46:34 jsonrpc2 handler: notification "workspace/didChangeConfiguration" handling error: json: cannot unmarshal string into Go struct field Config.settings.lintDebounce of type time.Duration
```

**Comment by ray-x** [comment]
Please take a look at https://github.com/golang/go/issues/10275
This is a known issue that `1s` in json can not be decoded to time.Duration

**Comment by mattn** [comment]
Now I understood what is wrong. Thank you.

**Comment by mattn** [comment]
Thank you

**Comment by ray-x** [comment]
Cheers!

**Comment by mattn** [comment]
Hmm, this break compatibility. Now efm-langserver can not load yaml configuration.

**Comment by ray-x** [comment]
What is the config.yaml setting breaks the loading?
I am using config.yaml with all settings copy from README.md. I did not see the issue

**Comment by mattn** [comment]
In previous version, efm-langserver does not parse below.

```
lint-debounce: 1s
```

Fixed in #147

**Comment by ray-x** [comment]
In fact. I think that part is missing in the config.yaml. Could you also update the document of how to setup lint-debounce/format-debouce in default config.yaml

**Pull request merged**